col*: use logical types throughout the vectorized engine

yuzefovich opened the pull request:
**col...: create new package and move some code**

This commit introduces `colbase` package which currently contains the
following (that are extracted from `colexec` package):
- `allocator.go`
- `random_testutils.go`
- `Operator` interface
- `CopyBatch` method.

The reason for extracting these things is that they are used by multiple
col* packages, so I think it's a good hygiene to separate them out.

This commit also renames `execerror` package to `vecerror` and then
moves it inside of `colbase` directory. It also changes the panic
matching so that now we catch all panics coming from a package that has
`pkg/sql/col` in its path (this will make sure that we don't forget to
register newly added packages that use panic-catching mechanism of the
vectorized engine with the panic-catcher). Additionally, this commit
renames the methods of `execerror` package.

Next, it moves `colexec/typeconv` package into `colbase` as well as
`colexec/testutils.go` file.

Finally, it removes `vecerror.NonVectorizedPanic` in favor of
`vecerror.ExpectedError`. The guidance on whether `InternalError` or
`ExpectedError` method should be used has been updated: the distinction
is whether the vectorized engine ends up in an unexpected - invalid -
state (for example, we expect that the vectorized engine might be
performing division by zero when evaluating a binary expression).

Release note: None

**col...: use logical types throughout the vectorized engine**

This commit transitions the vectorized engine to use SQL logical types
wherever possible, only converting to its physical type equivalent when
necessary (for example, when choosing which instance of projection
operator to use). This will allow us to have access to the actual type
whever we need and will allow us implement `coltypes.Datum` for all
unoptimized types.

This commit also moves some a few things around to clean up the
dependency graph (namely, `BytesEncodeFormat` is moved from
`sql/sessiondata` to `sql/lex`). Additionally, it replaces a single call
to `util/log.Infof` with a print statement in `util/protoutil` package
to remove the dependency on `util/log` (which depends on a bunch of
other things) so that `pkg/workload` would need to import less things.

Another thing worth calling out is the creation of copies of types in
`execplan` file to make sure that we don't override the input sync
specs.

Addresses: #43559.

Release note: None

**col...: introduce new package and more code movement**

Move `coldata/random_testutils.go` into newly created package
`coldatatestutils`.

Move contents of `colbase/random_testutils.go` into `coldatatestutils`
package.

Rename `colbase` to `colexecbase`. Also remove templated comments from
import sections of `_tmpl` files in favor of adding vars that remove
unused warnings (those templated comments work poorly when
moving/renaming the dependencies).

Rename `vecerror` to `colexecerror`.

Move `CopyBatch` from `colexecbase` into `coldatatestutils`. Also remove
memory accounting from `CopyBatch`.

Move `typeconv` from `colexecbase` into `coltypes` folder.

Move `coldata/vec_test.go` into `coldata_test` package to prevent an
import cycle. Also move one unit test from `coldata/vec_test.go` into
`coldata/bytes_test.go`.

Move `colexecbase/allocator.go` into newly created `colmem` package.

Release note: None

yuzefovich committed:
col*: create new package and move some code

This commit introduces `colbase` package which currently contains the
following (that are extracted from `colexec` package):
- `allocator.go`
- `random_testutils.go`
- `Operator` interface
- `CopyBatch` method.
The reason for extracting these things is that they are used by multiple
col* packages, so I think it's a good hygiene to separate them out.

This commit also renames `execerror` package to `vecerror` and then
moves it inside of `colbase` directory. It also changes the panic
matching so that now we catch all panics coming from a package that has
`pkg/sql/col` in its path (this will make sure that we don't forget to
register newly added packages that use panic-catching mechanism of the
vectorized engine with the panic-catcher). Additionally, this commit
renames the methods of `execerror` package.

Next, it moves `colexec/typeconv` package into `colbase` as well as
`colexec/testutils.go` file.

Finally, it removes `vecerror.NonVectorizedPanic` in favor of
`vecerror.ExpectedError`. The guidance on whether `InternalError` or
`ExpectedError` method should be used has been updated: the distinction
is whether the vectorized engine ends up in an unexpected - invalid
- state (for example, we expect that the vectorized engine might be
performing division by zero when evaluating a binary expression).

Release note: None

yuzefovich committed:
col*: use logical types throughout the vectorized engine

This commit transitions the vectorized engine to use SQL logical types
wherever possible, only converting to its physical type equivalent when
necessary (for example, when choosing which instance of projection
operator to use). This will allow us to have access to the actual type
whever we need and will allow us implement `coltypes.Datum` for all
unoptimized types. Physical types are stored only in few operators where
type conversion has a noticeable impact on the performance.

This commit also moves some a few things around to clean up the
dependency graph (namely, `BytesEncodeFormat` is moved from
`sql/sessiondata` to `sql/lex`). Additionally, it replaces a single call
to `util/log.Infof` with a print statement in `util/protoutil` package
to remove the dependency on `util/log` (which depends on a bunch of
other things) so that `pkg/workload` would need to import less things.

Another thing worth calling out is the creation of copies of types in
`execplan` file to make sure that we don't override the input sync
specs.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47583)
<!-- Reviewable:end -->

yuzefovich commented:
I ran the benchmarks of `colexec` package on gceworker between this branch and master, and the results are [here](https://gist.github.com/yuzefovich/a4743518fb73c280464401ffdfc12751). TL;DR is that there are no major changes in the run times, but there is slight increase in allocations in some cases. I'd say that doesn't concern me much. (Compare it to the initial benchmarks [here](https://gist.github.com/yuzefovich/58fc04a1822ed55ff3a186e47eeb9cfe) - those had some concerns which prompted me to store the converted physical types in a few places where constant conversion from logical to physical types matters.)

asubiotto reviewed (commented):
Great to see this. After we discussed, I was under the impression we wanted to remove `coltypes` completely. Looking at this code I don't think it'd be too hard to do to switch on the family, right? I recall you weren't sure you wanted to do this but let's discuss.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47583#-:-M5732p2EI3GiG9xq359:b6mfhqe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Copy pasting Alfonso's comments on the first four commits.

**Commit 1**:
The `colbase` package seems like it's just got a lot of things that don't belong in `colexec`. For example, the testing utilities seem like they can be moved to `coldata` since they're specifically about generating random data and you're probably already importing that if working with batches. Alternatively it should go into some `coldatatestutils` package or something. I think that should probably also include `CopyBatch` since it's used only in testing code and is supposed to be more of a testing utility anyway.

Regarding `Operator`, could you give an example of how it's beneficial to move it to a separate package? I don't doubt it because we do the same thing with the `Processor` interface in `execinfra`, but just want to wrap my head around it. Not sure what we should name this package. Maybe `colexecbase` is fine cc @jordanlewis 

The allocator seems like the thing we want to move out of `colexec` the most. Does it make sense to move it to `coldata` or is there some obstacle? Alternatively I think we should move it to its own package.

**Commit 2**:
I don't like the introduction of our first package (as far as I know) with the `vec` prefix if everything else is `col`. Maybe we should name more things `vec` (e.g.`vecexec`) but looks like `col` is here to stay. Should this be `colexecerror` instead?

**Commit 3**:
Can this be squashed into commit 1? The same `testutils` opinion applies here. Also why not move `typeconv` into `coltypes`?

yuzefovich committed:
col...: introduce new package and more code movement

Move `coldata/random_testutils.go` into newly created package
`coldatatestutils`.

Move contents of `colbase/random_testutils.go` into `coldatatestutils`
package.

Rename `colbase` to `colexecbase`. Also remove templated comments from
import sections of `_tmpl` files in favor of adding vars that remove
unused warnings (those templated comments work poorly when
moving/renaming the dependencies).

Rename `vecerror` to `colexecerror`.

Move `CopyBatch` from `colexecbase` into `coldatatestutils`. Also remove
memory accounting from `CopyBatch`.

Move `typeconv` from `colexecbase` into `coltypes` folder.

Move `coldata/vec_test.go` into `coldata_test` package to prevent an
import cycle. Also move one unit test from `coldata/vec_test.go` into
`coldata/bytes_test.go`.

Move `colexecbase/allocator.go` into newly created `colmem` package.

Release note: None

yuzefovich commented:
> The `colbase` package seems like it's just got a lot of things that don't belong in `colexec`. For example, the testing utilities seem like they can be moved to `coldata` since they're specifically about generating random data and you're probably already importing that if working with batches. Alternatively it should go into some `coldatatestutils` package or something. I think that should probably also include `CopyBatch` since it's used only in testing code and is supposed to be more of a testing utility anyway.

I like the idea of `coldatatestutils` package, done.

> Regarding `Operator`, could you give an example of how it's beneficial to move it to a separate package? I don't doubt it because we do the same thing with the `Processor` interface in `execinfra`, but just want to wrap my head around it.

The only reason for extracting `Operator` interface was existence `RandomDataOp` (which is now moved into `coldatatestutils`.

> Not sure what we should name this package. Maybe `colexecbase` is fine

Ok, I don't have any better idea for the name, renamed to `colexecbase`.

> The allocator seems like the thing we want to move out of `colexec` the most. Does it make sense to move it to `coldata` or is there some obstacle? Alternatively I think we should move it to its own package.

Indeed, moving `allocator` out of `colexec` was my main reason for creating a new package and it was the first thing I moved. Then I thought that I might as well extract other things that are used by multiple `col*` packages into that new package. I don't see any obstacle to moving `allocator.go` into `coldata`, but to me that seems a little off - `coldata` is responsible for providing an interface to store the data in columnar format whereas `allocator` concerns itself with memory accounting of those columnar allocations. I like having it in a separate package, and I think it is "closer" to the vectorized engine rather than the columnar format, so I'd leave in `colexecbase`.

> I don't like the introduction of our first package (as far as I know) with the `vec` prefix if everything else is `col`. Maybe we should name more things `vec` (e.g.`vecexec`) but looks like `col` is here to stay. Should this be `colexecerror` instead?

I agree with this sentiment and I like your idea of `colexecerror`. At the same time, I think we should decide to use either "columnar" or "vectorized" to reduce possible confusion among folks outside of our small team - it seems to me that we also speak of "vectorized engine" yet all our packages have "col" as the prefix. We can leave this discussion till later though.

> Can this be squashed into commit 1? The same `testutils` opinion applies here.

Initially I wanted to avoid conflicts due the rename of `execerror` to `vecerror` in between the commits you mentioned. Now all four commits are squashed.

>  Also why not move `typeconv` into `coltypes`?

I like this suggestion, done.

yuzefovich commented:
> Great to see this. After we discussed, I was under the impression we wanted to remove coltypes completely. Looking at this code I don't think it'd be too hard to do to switch on the family, right? I recall you weren't sure you wanted to do this but let's discuss.

I think currently there are two reasons for not fully removing `coltypes` and performing switch on the type families:
1. `IntFamily` is mapped to three physical types, so we would need to change the type-switching logic to accommodate that. The complication here is that we would need to look at the width of the type, so it would not be pure "family - switch", and more like "family - and maybe width of actual type - switch"
2. we have cases when multiple families map to the same physical type. If we are to do switch on the families, we will be generating more code that contains `case` blocks that are exact copies of each other.

I feel like we want to have the physical types whenever we're touching the physical representation of the logical type. However, coming up with strict guidelines on where the boundary between logical and physical should lie is hard.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47583)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/workload/dep_test.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47583#-M58MFOQCevvtKvBaS0X:-M58MFOQCevvtKvBaS0Y:b-bgtjlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/21fecec1f9e05082025beb4d73034f43375bf3b7/pkg/workload/dep_test.go#L29)):*
> ```Go
> 			`github.com/cockroachdb/cockroach/pkg/col/coldata`,
> 			`github.com/cockroachdb/cockroach/pkg/col/coltypes`,
> 			`github.com/cockroachdb/cockroach/pkg/col/coltypes/typeconv`,
> ```

cc @danhhz 

To provide some context for this PR: we want to introduce `coltypes.Datum` as a fallback for all types that are not explicitly supported by the vectorized engine (see #45612). This requires us plumbing `pkg/sql/types` into pretty much all packages in `pkg/col` folder, so `pkg/sql/types` now becomes a dependency for `pkg/workload` which brings all these new things into the dependencies as well.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
If we have only the allocator in `colexecbase`, I think I prefer naming it `colmem` and keeping only the allocator in there. I don't think there is a strong reason why I prefer this (I mostly just like when packages are specific), so up to you.

> I think currently there are two reasons for not fully removing coltypes and performing switch on the type families:
IntFamily is mapped to three physical types, so we would need to change the type-switching logic to accommodate that. The complication here is that we would need to look at the width of the type, so it would not be pure "family - switch", and more like "family - and maybe width of actual type - switch"

I don't think this is an issue. Switching on the width is ok and I think something we do in a lot of other places too.

> we have cases when multiple families map to the same physical type. If we are to do switch on the families, we will be generating more code that contains case blocks that are exact copies of each other.

Can you provide a concrete example of this? I'm comfortable with switching `case coltypes.Bytes:` to `case <family represented as bytes 1>, <family represented as bytes 2>`. I agree that it's easier to just do the former, but I don't think this justifies having and maintaining an extra type system.

> I feel like we want to have the physical types whenever we're touching the physical representation of the logical type.

Why? The `coltypes` are still an abstraction of a physical type. This might be a naive statement but I think the only reason they exist is to map types to their physical representations. I think this mapping can already be done by switching on logical type families and working with the actual physical type representations. i.e. the only purpose of `coltypes` is a translation, and since we can do this translation with logical types, why keep `coltypes` around?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47583#-:-M5MBwEh0-gZttRmnnCH:bul3asf)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
> If we have only the allocator in `colexecbase`, I think I prefer naming it `colmem` and keeping only the allocator in there. I don't think there is a strong reason why I prefer this (I mostly just like when packages are specific), so up to you.

I extracted `colmem` package which currently only contains `allocator.go` file, but I think we still want to keep `colexecbase` package.

As discussed, I'll be addressing other comments in a follow-up PR.

asubiotto commented on pkg/col/coldatatestutils/random_testutils.go:
```diff
@@ -110,14 +229,14 @@ const (
 // optional (refer to the constants above this struct definition for the
 // defaults). Bools are false by default and AvailableTyps defaults to
-// coltypes.AllTypes.
+// colexecbase.AllSupportedSQLTypes.
```
This should be typeconv.AllSupportedSQLTypes, right?

asubiotto commented on pkg/sql/colcontainer/diskqueue_test.go:
```diff
@@ -199,9 +200,9 @@ func BenchmarkDiskQueue(b *testing.B) {
-	op := colexec.NewRepeatableBatchSource(testAllocator, batch)
+	typs := []types.T{*types.Int}
+	batch := coldatatestutils.RandomBatch(testAllocator, rng, typs, coldata.BatchSize(), 0, 0)
+	op := colexecbase.NewRepeatableBatchSource(testAllocator, batch, typs)
```
Does RepeatableBatchSource belong in `colexecbase` or a `testutils` package?

asubiotto reviewed (approved):
LGTM

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47583#-:-M5SBq-dCXqhHM0kYBpB:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colcontainer/diskqueue_test.go, line 205 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47583#-M53nMSpqXcSz1CKHxql-r6-205:-M5SB443BJgN2yah36-9:b-3clmuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cc3559291c226edc6df54cfb67e6fe72d436cc7/pkg/sql/colcontainer/diskqueue_test.go#L205)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Does RepeatableBatchSource belong in `colexecbase` or a `testutils` package?
</blockquote></details>

I don't think it belongs in `coldatatestutils` because the struct is a `colexecbase.Operator`. We could add another tiny package `sql/coltestutils` (which would depend on `colexecbase`) or something like that, but then we would have 3 new packages in this PR with one file in each, and personally I don't like having too many packages without clear benefits.
___
*[pkg/col/coldatatestutils/random_testutils.go, line 232 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47583#-M58J034Xa4mRNYK_64K-r6-232:-M5SA5Vp2tW12YQS11Mx:b-pkzycw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cc3559291c226edc6df54cfb67e6fe72d436cc7/pkg/col/coldatatestutils/random_testutils.go#L232)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This should be typeconv.AllSupportedSQLTypes, right?
</blockquote></details>

Yes, fixed.


<!-- Sent from Reviewable.io -->
